config/hotreloadable.go

```diff
@@ -1,14 +1,56 @@
 package config
 
-import "time"
+import (
+	"sync"
+	"time"
+)
 
 // RegisterIntConfigVariable registers int config variable
+// Deprecated: use RegisterIntVar or RegisterAtomicIntVar instead
 func RegisterIntConfigVariable(defaultValue int, ptr *int, isHotReloadable bool, valueScale int, keys ...string) {
 	Default.RegisterIntConfigVariable(defaultValue, ptr, isHotReloadable, valueScale, keys...)
 }
 
+// RegisterIntVar registers a not hot-reloadable int config variable
+func RegisterIntVar(defaultValue int, ptr *int, valueScale int, keys ...string) {
+	Default.RegisterIntVar(defaultValue, ptr, valueScale, keys...)
+}
+
+// RegisterAtomicIntVar registers a hot-reloadable int config variable
+func RegisterAtomicIntVar(defaultValue, valueScale int, keys ...string) *Atomic[int] {
+	return Default.RegisterAtomicIntVar(defaultValue, valueScale, keys...)
+}
+
 // RegisterIntConfigVariable registers int config variable
-func (c *Config) RegisterIntConfigVariable(defaultValue int, ptr *int, isHotReloadable bool, valueScale int, keys ...string) {
+// Deprecated: use RegisterIntVar or RegisterAtomicIntVar instead
+func (c *Config) RegisterIntConfigVariable(
+	defaultValue int, ptr *int, isHotReloadable bool, valueScale int, keys ...string,
+) {
+	c.registerIntVar(defaultValue, ptr, isHotReloadable, valueScale, func(v int) {
+		*ptr = v
+	}, keys...)
+}
+
+// RegisterIntVar registers a not hot-reloadable int config variable
+func (c *Config) RegisterIntVar(defaultValue int, ptr *int, valueScale int, keys ...string) {
+	c.registerIntVar(defaultValue, ptr, false, valueScale, func(v int) {
+		*ptr = v
+	}, keys...)
+}
+
+// RegisterAtomicIntVar registers a hot-reloadable int config variable
+// Copy of RegisterIntConfigVariable, but with a way to avoid data races for hot reloadable config variables
+func (c *Config) RegisterAtomicIntVar(defaultValue, valueScale int, keys ...string) *Atomic[int] {
+	ptr := getOrCreatePointer(c.atomicVars, c.atomicVarsMisuses, &c.atomicVarsLock, defaultValue, keys...)
```

Comment:
Whenever we are reusing an existing pointer we shouldn't register it again. 
If `getOrCreatePointer` returned a second value indicating whether this is a new or existing pointer, we could use this information for avoiding registering it twice.

One thing we should definitely fix though is whenever we are calling `appendVarToConfigMaps`, we should pass as a key the **full concatenated key** instead of the first key from the slice.

Reply:
> One thing we should definitely fix though is whenever we are calling `appendVarToConfigMaps`, we should pass as a key the **full concatenated key** instead of the first key from the slice.

Nice, good catch :+1: 

> Whenever we are reusing an existing pointer we shouldn't register it again. If `getOrCreatePointer` returned a second value indicating whether this is a new or existing pointer, we could use this information for avoiding registering it twice.

We could allow this though, it is now safe to do so and you'll get the same variable. I'm keen to let people do it at this point. If someone misuses the default values will get a panic as well. Perhaps we should settle on the order of the keys too, I might be missing some use case.